In [1]:
import pandas as pd
import numpy as np
import pickle

from crossref.restful import Works, Etiquette
#from scihub_blastoise import SciHub_HydroPump


In [2]:
with open("q1_jrnl_expladed", "rb") as fp:
    q1_jrnl_expladed = pickle.load(fp)

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

__NOTE:__ Before the loop, it is realised that using q1_jrnl_df during the loop will prevent duplicate article extraction by a HUGE margin. therefore, although it will be double-work, it makes much more sense to use it & attach s_a info to articles after the extraction is completed!

TTD:

* Need to convert ISSN colm to list - DONE - in SumCetCet below
* Need to update useful_cols - DONE - only removed "reference" column
* In addition to prob.13, there are 44 journals that exceed their CR_Count when the Current_Count is also considered. Fortunately, this is also a very low minority when the total dataset is considered, but still neeeds to be accounted for! - SKIPPED - will deal with later

In [37]:
# TTD #2:
with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

print(q1_date_df.columns)

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


Index(['JRNL_ID', 'ISSN', 'issued', 'container-title', 'publisher', 'title',
       'author', 'subject', 'references-count', 'is-referenced-by-count',
       'published', 'published-print', 'published-online', 'direct_url',
       'Dates', 'Keywords', 'Affiliation'],
      dtype='object')


In [ ]:
# TTD #3:
q1_jrnl_df[q1_jrnl_df.Required_Count + q1_jrnl_df.Current_Count > q1_jrnl_df.CR_Count]

In [30]:
# Test Run:

# SumCetCet CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette

testo = q1_jrnl_fe.sample(5)
print(testo.Required_Count.sum())


my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', 'oguzkokes@gmail.com')

ogi_works = Works(etiquette=my_etiquette)

q1_dict = testo[["ISSN","Year", "Required_Count"]].to_dict("index")

#Need to convert ISSN column to list!!!


cet_q1_meta_fe = []
useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=list(issn), from_pub_date=year, until_pub_date=year, type="journal-article").sample(int(rc)).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    cet_q1_meta_fe.extend(loop_meta)


# open("cet_q1_meta_fe","wb") as fp:
#    pickle.dump(cet_q1_meta_fe, fp)

Test run is succesful! Will move on to the SumCetCet run!

In [ ]:
# SumCetCet CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette
import pandas as pd
import pickle

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

# Filtering only necessary jrnls:
q1_jrnl_fe = q1_jrnl_df[q1_jrnl_df.Required_Count >0].copy()
q1_jrnl_fe["ISSN"] = q1_jrnl_fe.ISSN.map(list)

#SumCetCet division:
q1_cr_sum = q1_jrnl_fe[:7000]
q1_cr_cet = q1_jrnl_fe[70000:170000]
q1_cr_ogi = q1_jrnl_fe[170000:]

mail = ""
# 'sduzenli@yandex.com.tr'
# 'cetiner.mserdar@gmail.com'
# 'oguzkokes@gmail.com'

my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)


ogi_works = Works(etiquette=my_etiquette)

q1_dict = q1_cr_cet[["ISSN","Year", "Required_Count"]].to_dict("index")


cet_q1_meta_fe = []
useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']


for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    cet_q1_meta_fe.extend(loop_meta)


with open("cet_q1_meta_fe","wb") as fp:
    pickle.dump(cet_q1_meta_fe, fp)

In [1]:
# Ogi CR Loop Run (for q1_jrnl_fe):

from crossref.restful import Works, Etiquette
import pandas as pd
import pickle

with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

# Filtering only necessary jrnls:
q1_jrnl_fe = q1_jrnl_df[q1_jrnl_df.Required_Count >0].copy()
q1_jrnl_fe["ISSN"] = q1_jrnl_fe.ISSN.map(list)

#SumCetCet division:
q1_cr_ogi = q1_jrnl_fe[17000:]


mail = 'oguzkokes@gmail.com'
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', mail)
ogi_works = Works(etiquette=my_etiquette)

q1_dict = q1_cr_ogi[["ISSN","Year", "Required_Count"]].to_dict("index")


ogi_q1_meta_fe = []

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    ogi_q1_meta_fe.extend(loop_meta)


with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

An error occured after JRNL 1_49971, will continue loop from this point on!

In [9]:
# DUMPing ogi_q1_meta_fe just in case:

with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

In [11]:
# Cont. after error:

q1_dict = q1_cr_ogi.loc["1_49971":,["ISSN","Year", "Required_Count"]].to_dict("index")

useful_cols = ['DOI', 'references-count', 'publisher', 'published-print', 'is-referenced-by-count', 'title', 'author', 'published-online', 'container-title', 'issued', 'ISSN', 'subject']

for key in q1_dict:
    issn = q1_dict[key]["ISSN"]
    year = q1_dict[key]["Year"]
    rc = q1_dict[key]["Required_Count"]

    loop_meta = [w for w in ogi_works.filter(issn=issn, from_pub_date=year, until_pub_date=year, type="journal-article").sample(rc).select(useful_cols)] 
    loop_meta = [dict(item, JRNL_ID=key) for item in loop_meta]
        
    ogi_q1_meta_fe.extend(loop_meta)


with open("ogi_q1_meta_fe","wb") as fp:
    pickle.dump(ogi_q1_meta_fe, fp)

NOTE: The "problematic" journal is duplicated, so need to drop half the extracred artlcs after the run is completed!

In [21]:
q1_jrnl_fe[17000:].Required_Count.sum()

42992

## Part 2: After SumCetCet FE - CrossRef Run

SumCetCet run is complete will now load all fe dfs & analyze results:

In [23]:
with open("sum_q1_meta_fe", "rb") as fp:
    sum_q1_meta_fe = pickle.load(fp)

with open("cet_q1_meta_fe", "rb") as fp:
    cet_q1_meta_fe = pickle.load(fp)

# with open("cet_q1_meta_fe", "rb") as fp:
#     cet_q1_meta_fe = pickle.load(fp)

In [31]:
q1_cr_sum = q1_jrnl_fe[:7000]
q1_cr_cet = q1_jrnl_fe[7000:17000]
# q1_cr_ogi = q1_jrnl_fe[170000:]



In [29]:
# Sum Analysis:

print(f"Required Count: {q1_cr_sum.Required_Count.sum()}")

print(f"Extracted Count: {len(sum_q1_meta_fe)}")

Required Count: 41367
Extracted Count: 41367


In [33]:
# Cet Analysis:

print(f"Required Count: {q1_cr_cet.Required_Count.sum()}")

print(f"Extracted Count: {len(cet_q1_meta_fe)}")

print(f"Difference: {53907-53836}")

Required Count: 53907
Extracted Count: 53836
Difference: 71


In [35]:
# Ogi Analysis:

print(f"Required Count: {q1_cr_ogi.Required_Count.sum()}")

print(f"Extracted Count: {len(ogi_q1_meta_fe)}")

print(f"Difference: {42992-42981}")

Required Count: 42992
Extracted Count: 42981
Difference: 11


In [40]:
q1_meta_fe = pd.concat([pd.DataFrame(sum_q1_meta_fe),pd.DataFrame(cet_q1_meta_fe),pd.DataFrame(ogi_q1_meta_fe)], ignore_index=True)

In [43]:
q1_meta_fe.head()

publisher              published-print                DOI  \
0     Wiley  {'date-parts': [[2013, 6]]}  10.1002/aic.13989   
1     Wiley  {'date-parts': [[2014, 1]]}  10.1002/aic.14236   
2     Wiley  {'date-parts': [[2013, 2]]}  10.1002/aic.13983   
3     Wiley  {'date-parts': [[2013, 6]]}  10.1002/aic.14123   
4     Wiley  {'date-parts': [[2014, 1]]}  10.1002/aic.14283   

   is-referenced-by-count                                              title  \
0                       9  [Liquid Fuels from Alternative Carbon Sources ...   
1                      29  [Microwave irradiated and thermally heated oli...   
2                       6  [Efficiency of Nitrogen Desorption from LiX Ze...   
3                       3  [Celebrating singularities: Mathematics and ch...   
4                      12  [Co-firing technology for the preparation of a...   

                                              author  \
0  [{'given': 'Matthew J.', 'family': 'Metzger', ...   
1  [{'given': 'Tamer M.', 'family': 'Alslaibi', '...   
2  [{'given': 'Siew Wah', 'family': 'Chai', 'sequ...   
3  [{'given': 'Yunshan', 'family': 'Wang', 'seque...   
4  [{'given': 'U.', 'family': 'Pippardt', 'sequen...   

                   published-online  container-title  \
0   {'date-parts': [[2013, 2, 19]]}  [AIChE Journal]   
1   {'date-parts': [[2013, 10, 8]]}  [AIChE Journal]   
2    {'date-parts': [[2013, 1, 3]]}  [AIChE Journal]   
3    {'date-parts': [[2013, 5, 8]]}  [AIChE Journal]   
4  {'date-parts': [[2013, 11, 19]]}  [AIChE Journal]   

                             issued  references-count         ISSN  \
0   {'date-parts': [[2013, 2, 19]]}                50  [0001-1541]   
1   {'date-parts': [[2013, 10, 8]]}                52  [0001-1541]   
2    {'date-parts': [[2013, 1, 3]]}                10  [0001-1541]   
3    {'date-parts': [[2013, 5, 8]]}                62  [0001-1541]   
4  {'date-parts': [[2013, 11, 19]]}                20  [0001-1541]   

                                             subject JRNL_ID  
0  [General Chemical Engineering, Environmental E...     1_0  
1  [General Chemical Engineering, Environmental E...     1_0  
2  [General Chemical Engineering, Environmental E...     1_0  
3  [General Chemical Engineering, Environmental E...     1_0  
4  [General Chemical Engineering, Environmental E...     1_0

In [44]:
with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

In [49]:
q1_date_df.sample(10)

JRNL_ID                    ISSN  \
DOI                                                              
10.3389/feart.2020.574206      1_65355             [2296-6463]   
10.1111/arcm.12572               1_680  (0003-813X, 1475-4754)   
10.1016/j.jsbmb.2015.01.009    1_20371             [0960-0760]   
10.1080/14712598.2018.1544239  1_36613  [1471-2598, 1744-7682]   
10.1115/1.4038879              1_17283  [0889-504X, 1528-8900]   
10.1557/mrc.2014.14            1_63492  [2159-6859, 2159-6867]   
10.1111/sapm.12220             1_34849  [0022-2526, 1467-9590]   
10.1016/j.ejor.2010.04.017     1_15010             [0377-2217]   
10.1109/tcsi.2010.2055250      1_44525  (1549-8328, 1558-0806)   
10.1007/s13369-015-1920-6      1_64142  [1319-8025, 2191-4281]   

                                                         issued  \
DOI                                                               
10.3389/feart.2020.574206        {'date-parts': [[2020, 9, 2]]}   
10.1111/arcm.12572               {'date-parts': [[2020, 5, 8]]}   
10.1016/j.jsbmb.2015.01.009         {'date-parts': [[2015, 4]]}   
10.1080/14712598.2018.1544239  {'date-parts': [[2018, 11, 14]]}   
10.1115/1.4038879                {'date-parts': [[2018, 4, 6]]}   
10.1557/mrc.2014.14                 {'date-parts': [[2014, 6]]}   
10.1111/sapm.12220              {'date-parts': [[2018, 6, 25]]}   
10.1016/j.ejor.2010.04.017         {'date-parts': [[2010, 11]]}   
10.1109/tcsi.2010.2055250           {'date-parts': [[2011, 1]]}   
10.1007/s13369-015-1920-6      {'date-parts': [[2015, 10, 27]]}   

                                                                 container-title  \
DOI                                                                                
10.3389/feart.2020.574206                           [Frontiers in Earth Science]   
10.1111/arcm.12572                                                [Archaeometry]   
10.1016/j.jsbmb.2015.01.009    [The Journal of Steroid Biochemistry and Molec...   
10.1080/14712598.2018.1544239             [Expert Opinion on Biological Therapy]   
10.1115/1.4038879                                    [Journal of Turbomachinery]   
10.1557/mrc.2014.14                                         [MRS Communications]   
10.1111/sapm.12220                              [Studies in Applied Mathematics]   
10.1016/j.ejor.2010.04.017            [European Journal of Operational Research]   
10.1109/tcsi.2010.2055250      [IEEE Transactions on Circuits and Systems I: ...   
10.1007/s13369-015-1920-6          [Arabian Journal for Science and Engineering]   

                                                                       publisher  \
DOI                                                                                
10.3389/feart.2020.574206                                     Frontiers Media SA   
10.1111/arcm.12572                                                         Wiley   
10.1016/j.jsbmb.2015.01.009                                          Elsevier BV   
10.1080/14712598.2018.1544239                                 Informa UK Limited   
10.1115/1.4038879                                             ASME International   
10.1557/mrc.2014.14                      Springer Science and Business Media LLC   
10.1111/sapm.12220                                                         Wiley   
10.1016/j.ejor.2010.04.017                                           Elsevier BV   
10.1109/tcsi.2010.2055250      Institute of Electrical and Electronics Engine...   
10.1007/s13369-015-1920-6                Springer Science and Business Media LLC   

                                                                           title  \
DOI                                                                                
10.3389/feart.2020.574206      [Temperature Variation on the Central Tibetan ...   
10.1111/arcm.12572             [Evaluation of Fe isotope values as a provenan...   
10.1016/j.jsbmb.2015.01.009    [The effect of 1,25 dihydroxyvitamin D3 treatm..

Here, it is realized that the S_A column  for q1_date_df was created in (7a) but it was not saved. 

This isnot a huge loss, as we can assign the S_A after the fe SH_run & the entire dataset is collected. This can also allow a better assignment regard Currecnt_Count & Required_Count & CR_Count colunms.

---

The next step is SH_run for fe dataset!

In [53]:
with open("q1_meta_fe","wb") as fp:
    pickle.dump(q1_meta_fe, fp)

In [54]:
q1_meta_fe.shape

(138184, 13)

In [ ]:
# q1_meta_fe SH_Blastoise Ogi Run:

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[95000:]


doi_list = q1_meta_fe_ogi.DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [62]:
# Test Run:

doi_list = q1_meta_fe.sample(5).DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


In [66]:
pd.DataFrame(q1_date_fe_ogi)

DOI  \
0     10.1088/0268-1242/30/8/084001   
1         10.1017/s0954579419000130   
2          10.1177/1074248413504994   
3         10.1016/j.whi.2012.10.001   
4  10.1111/j.1755-263x.2010.00162.x   

                                          direct_url  \
0  https://sci.bban.top/pdf/10.1088/0268-1242%252...   
1  https://sci.bban.top/pdf/10.1017/s095457941900...   
2  https://sci.bban.top/pdf/10.1177/1074248413504...   
3  https://sci.bban.top/pdf/10.1016/j.whi.2012.10...   
4  https://sci.bban.top/pdf/10.1111/j.1755-263x.2...   

                                Dates  \
0                 [31 May 2015, None]   
1                       no_date_found   
2     [May 28, 2013, August 15, 2013]   
3       [8 June 2012, 9 October 2012]   
4  [7 October 2010, 16 December 2010]   

                                            Keywords  
0  [(igbts, 0.3377), (mosfets, 0.3217), (igbt, 0....  
1  [(psychoneuroendocrinology, 0.4846), (stressor...  
2  [(atherosclerosis, 0.5629), (atherothrombosis,...  
3  [(contraception, 0.6124), (contraceptive, 0.56...  
4  [(wildlife, 0.6136), (tiger, 0.5459), (tigers,...

Test run is completed! SumCetCet run can be initiated!

As the whole loop might take a while, will start the loop in chunks below:

__NOTE: DON'T FORGET TO KEEP q1_date_fe_ogi & DO NOT RE-CREATE!__

In [73]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 1

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[95000:105000]


doi_list = q1_meta_fe_ogi.DOI.tolist()

q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi_part1_7c","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [84]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 2

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[105000:120000]


doi_list = q1_meta_fe_ogi.DOI.tolist()

#q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi_part2_7c","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


In [86]:
# q1_meta_fe SH_Blastoise Ogi Run: - PART 3

# from scihub_blastoise import SciHub_HydroPump
# import pickle

sh_h = SciHub_HydroPump()

# with open("q1_meta_fe", "rb") as fp:
#     q1_meta_fe = pickle.load(fp)

#q1_meta_fe_sum = q1_meta_fe[:45000]
#q1_meta_fe_cet = q1_meta_fe[45000:95000]
q1_meta_fe_ogi = q1_meta_fe[120000:]


doi_list = q1_meta_fe_ogi.DOI.tolist()

#q1_date_fe_ogi = list() 


for doi in doi_list:
    pdf_url, date, kw = sh_h.get_dates_kws(doi)

    q1_date_fe_ogi.append(dict(DOI = doi, direct_url = pdf_url, Dates = date, Keywords = kw))


with open("q1_date_fe_ogi","wb") as fp:
    pickle.dump(q1_date_fe_ogi, fp)


## Part 3: After SciHub Run:

SumCetCet for Q1_fe is completed! The nb will be reset and results will be analyzed.

Depending on the extracted information, will re-calculate Required_Count & update nb info accodringly!

In [2]:
with open("q1_date_fe_sum", "rb") as fp:
    q1_date_fe_sum = pickle.load(fp)

with open("q1_date_fe_cet", "rb") as fp:
    q1_date_fe_cet = pickle.load(fp)

with open("q1_date_fe_ogi", "rb") as fp:
    q1_date_fe_ogi = pickle.load(fp)

In [3]:
q1_date_fe_sum = pd.DataFrame(q1_date_fe_sum)
q1_date_fe_cet = pd.DataFrame(q1_date_fe_cet)
q1_date_fe_ogi = pd.DataFrame(q1_date_fe_ogi)

q1_date_fe = pd.concat([q1_date_fe_sum,q1_date_fe_cet,q1_date_fe_ogi], ignore_index=True)

In [17]:
q1_date_fe.Dates.map(type).value_counts()

<class 'list'>    108345
<class 'str'>      29839
Name: Dates, dtype: int64

In [18]:
q1_date_fe[q1_date_fe.Dates.map(type)==str].Dates.value_counts()

no_date_found            26748
article_not_in_Scihub     2951
cant_read_pdf              137
direct_url_error             2
pdf_bytes_error              1
Name: Dates, dtype: int64

It is seen that the majority of the errors are caused by "NO DATE". However, to clearly understand this issue, we need JRNL_ID column as well.

In [19]:
with open("q1_meta_fe", "rb") as fp:
    q1_meta_fe = pickle.load(fp)

q1_date_fe = q1_date_fe.merge(q1_meta_fe[["DOI","JRNL_ID"]], how="left", left_on="DOI", right_on="DOI")

In [33]:
q1_anlys = q1_date_fe.groupby("JRNL_ID").Dates.apply(list)
q1_anlys = pd.DataFrame(q1_anlys)

q1_anlys["list_total"] = q1_anlys.Dates.map(len)
q1_anlys["list_succ"] = q1_date_fe.groupby("JRNL_ID").Dates.apply(list).map(lambda x: len([i for i in x if type(i)==list]))
q1_anlys["list_diff"] = q1_anlys["list_total"] - q1_anlys["list_succ"]

In [79]:
# PARTIAL OK:

len(q1_anlys[q1_anlys.list_diff > 0].sort_values("list_diff",ascending=False))

11054

In [78]:
#FULL OK:
len(q1_anlys[q1_anlys.list_diff == 0])

14012

In [80]:
# ZERO OK:
len(len(q1_anlys[q1_anlys.list_succ == 0].sort_values("list_total", ascending=False)))

1976

There are 3 types of jrnls:

- _FULL OK_ : These journals have completely satistfied the Required_count conditions. These jrnls might be used to instead of failed retr. 

- _PARTIAL OK_ : These journals have DATE INFO only for SOME jrnls. These jrnls should be subjected to a custom loop to get maximum artcls

- _ZERO OK_ : These jrnls did not have any DATE INFO from extracted artcls although they were marked as SH_Results == True

To get a better understanding of the situation & calc & create new loops, we need q1_jrnl_df imported

In [48]:
with open("q1_jrnl_df", "rb") as fp:
    q1_jrnl_df = pickle.load(fp)

with open("q1_jrnl_expladed", "rb") as fp:
    q1_jrnl_expladed = pickle.load(fp)

In [51]:
q1_jrnl_df = q1_jrnl_df.merge(q1_anlys[["list_total","list_succ","list_diff"]], how="left", right_index=True, left_index=True)

In [90]:
# NOT in CrossRef:
print(f"Number of not included in CR journals: {len(q1_jrnl_df[(q1_jrnl_df.list_total.isna()) & (q1_jrnl_df.Required_Count > 0)])}")


print(f'Number of artcls from these journals: {q1_jrnl_df[(q1_jrnl_df.list_total.isna()) & (q1_jrnl_df.Required_Count > 0)]["Required_Count"].sum()}')




Number of not included in CR journals: 9
Number of artcls from these journals: 78


In [82]:
# CR_Count changed jrnls:

len(q1_jrnl_df[(q1_jrnl_df.Required_Count > q1_jrnl_df.list_total)])

1

In [77]:
my_etiquette = Etiquette('Analysing Publishing Delay in Academic Journals', 'v2.0', 'https://github.com/Spidey0023/THEsis-Codes', 'oguzkokes@gmail.com')
ogi_works = Works(etiquette=my_etiquette)


ogi_works.filter(issn=["10870571", "1552454X"], from_pub_date=2016, until_pub_date=2016, type="journal-article").count() 

1

__The analysis so far shows that:__

We have 2 areas to focus:

___JRNL related issues: (CrossRef)___

- 9 jrnls were removed from CrossRef entirely!

- 1 jrnl, 1_24843, had CR_Count 78, but now only 1 


___ARTCL related issues: (NO DATE & not in SH)___

- 1976 jrnls were marked as SH_Results = True, but had 0 successful retrial => complete failure - ZERO_OK

- 11054 jrnls PARTIAL_OK

- rest worked completely as expected!


---

___TTD___:

1- create new version of q1_date_df,

- save q1_date_fe
- save q1_meta_fe
- filter & inner join together
- modify & feat. eng 
- merge with q1_date_df to create new version!

2- q1_jrnl_df update:

- 9+1 CR_Count, 
    
- 1976 SH_Results

3- def partial func

4- re-check q1_jrnl_df for CR_Count & SH_Results?

#### 1 - q1_date_df updates:

In [92]:
#1 q1_date_fe & q1_date_df merging:

# 1.1 - save q1_date_fe:
q1_date_fe_all = q1_date_fe.copy()

with open("q1_date_fe_all","wb") as fp:
   pickle.dump(q1_date_fe_all, fp)

# Saved in 7c SH run folder!

# Also duplicated q1_meta_fe_all & SAVED in same folder!

In [104]:
# 1.2 - Merging the two:
q1_date_fe = q1_meta_fe.merge(q1_date_fe.drop("JRNL_ID", axis=1), how="inner", left_on="DOI", right_on="DOI")

# 1.3 - Filtering to only successful artcls:
q1_date_fe = q1_date_fe[q1_date_fe.Dates.map(type)==list]
q1_date_fe.set_index("DOI", drop=True, inplace=True)

In [110]:
# 1.4 - Feature Eng. : 
def earlier_date_part(row):
    try:
        pp_year = row["published-print"]["date-parts"][0][0]
    except TypeError:
        pp_year = None

    try:   
        po_year = row["published-online"]["date-parts"][0][0]
    except TypeError:
        po_year = None

    if po_year is None:
        return row["published-print"]

    elif pp_year is None:
        return row["published-online"]

    else:
        if pp_year<po_year:
            return row["published-print"]
        elif po_year<pp_year:
            return row["published-online"]
        else:
            try:
                pp_month = row["published-print"]["date-parts"][0][1]
            except IndexError:
                pp_month = 12
            try:
                po_month = row["published-online"]["date-parts"][0][1]
            except IndexError:
                po_month = 12
            if pp_month<po_month:
                return row["published-print"]
            elif po_month<pp_month:
                return row["published-online"]
            else:
                return row["published-online"]


q1_date_fe.loc[:,"published"] = q1_date_fe.apply(earlier_date_part, axis=1)

q1_date_fe = q1_date_fe[["JRNL_ID", "ISSN", "issued", "container-title", "publisher", "title", "author", "subject", "references-count", "is-referenced-by-count", "published", "published-print", "published-online", "direct_url", "Dates", "Keywords"]]


In [116]:
# 1.5 - Merge:

with open("q1_date_df", "rb") as fp:
    q1_date_df = pickle.load(fp)

It is seen that 1671 articles are duplicated btw q1_date_df & q1_jrnl_fe. When analyzed, the two have different values on "Dates" and "Keywords" columns (as expected). As the newer results are to be more precise & old 1671 will be dropped from q1_date_df

__Before the merge, q1_date_df is SAVED as q1_date_df_old in 7c SH folder!!!__

In [122]:
shared_dois = list(set(q1_date_df.index).intersection(set(q1_date_fe.index)))

In [154]:
# Proof that 2 dfs have different "Dates" for duplicate DOIs:
q1_date_df.loc[shared_dois,"Dates"].equals(q1_date_fe.loc[shared_dois,"Dates"])

False

In [158]:
q1_date_df = pd.concat([q1_date_df.drop(shared_dois), q1_date_fe], verify_integrity=True)

In [161]:
# 1.6 - Saving & DUMPing the updated q1_date_df:

with open("q1_date_df","wb") as fp:
   pickle.dump(q1_date_df, fp)

With this process, q1_date_df is up-to-date & ready for further steps.

However, q1_jrnl_df needs to be analyzed & updated (if necessary) as well...

#### 2 - q1_jrnl_df updates:

The initial analysis above showed that there were 3 different problems during q1_date_fe run. These are:

- for 9+1 jrnls, the CR Count is either no longer valid OR all artcls have been removed (CR_Count = 0)
    
    _all these jrnls had >1 Req_count meaning there is reason for a re-trial & these jnrls can be marked as CR_Count = 0 & 1 accordingly_


- for 1976 jrnls, there are no succesful DATE INFO extracted. Some of these jrnls such as 1_54942 had 37 artcls extracted & non had DATE INFO, so it can / should be dropped!
    
    _However, further analysis below show that most of these jrnls had only <3 artcls retrieved, and therefore should be re-tried!!!_

- for 11054 jrnls, the current status is PARTIAL OK, need to fix!

In [165]:
q1_jrnl_df.fillna(0)[q1_jrnl_df.fillna(0).Required_Count > q1_jrnl_df.fillna(0).list_total]

ISSN  Year                              Title  \
ID                                                                       
1_18052           (09096396,)  2013                   Wildlife Biology   
1_24838  (10870571, 1552454X)  2010  Journal of Biomolecular Screening   
1_24839  (10870571, 1552454X)  2012  Journal of Biomolecular Screening   
1_24840  (10870571, 1552454X)  2014  Journal of Biomolecular Screening   
1_24841  (10870571, 1552454X)  2011  Journal of Biomolecular Screening   
1_24842  (10870571, 1552454X)  2015  Journal of Biomolecular Screening   
1_24843  (10870571, 1552454X)  2016  Journal of Biomolecular Screening   
1_24844  (10870571, 1552454X)  2013  Journal of Biomolecular Screening   
1_66638  (24726311, 24726303)  2019                    SLAS Technology   
1_66639  (24726311, 24726303)  2020                    SLAS Technology   

                                                  Sbj_Area  Total_Docs  \
ID                                                                       
1_18052                       {'Environmental Science': 1}          51   
1_24838  {'Pharmacology, Toxicology and Pharmaceutics': 1}         132   
1_24839  {'Pharmacology, Toxicology and Pharmaceutics': 1}         139   
1_24840  {'Chemistry': 1, 'Biochemistry, Genetics and M...         139   
1_24841  {'Pharmacology, Toxicology and Pharmaceutics': 1}         129   
1_24842  {'Chemistry': 1, 'Biochemistry, Genetics and M...         125   
1_24843  {'Chemistry': 1, 'Biochemistry, Genetics and M...         113   
1_24844  {'Chemistry': 1, 'Pharmacology, Toxicology and...         118   
1_66638                          {'Health Professions': 1}          66   
1_66639                          {'Health Professions': 1}          73   

               Country            Region  \
ID                                         
1_18052        Denmark    Western Europe   
1_24838  United States  Northern America   
1_24839  United States  Northern America   
1_24840  United States  Northern America   
1_24841  United States  Northern America   
1_24842  United States  Northern America   
1_24843  United States  Northern America   
1_24844  United States  Northern America   
1_66638  United States  Northern America   
1_66639  United States  Northern America   

                                    Publisher   Coverage    SJR  ...  \
ID                                                               ...   
1_18052  Nordic Council for Wildlife Research  1995-2020  0,822  ...   
1_24838                SAGE Publications Inc.  1996-2017  1,017  ...   
1_24839                SAGE Publications Inc.  1996-2017  0,964  ...   
1_24840                SAGE Publications Inc.  1996-2017  1,155  ...   
1_24841                SAGE Publications Inc.  1996-2017  0,987  ...   
1_24842                SAGE Publications Inc.  1996-2017  1,080  ...   
1_24843                SAGE Publications Inc.  1996-2017  1,347  ...   
1_24844                SAGE Publications Inc.  1996-2017  1,000  ...   
1_66638                SAGE Publications Inc.  2017-2020  0,632  ...   
1_66639                SAGE Publications Inc.  2017-2020  0,714  ...   

         Citable Docs. (3years)  Cites / Doc. (2years)  Ref. / Doc.  CR_Count  \
ID                                                                              
1_18052                     130                   1,32        47,47        56   
1_24838                     318                   2,80        26,29       150   
1_24839                     373                   2,43        30,27       155   
1_24840                     381                   2,41        33,05       148   
1_24841                     341                   2,23        27,72       169   
1_24842                     393                   2,47        39,44       111   
1_24843                     379                   2,54        31,75        78   
1_24844                     394                   2,11        32,87       141   
1_66638                     125            

In [151]:
q1_anlys[q1_anlys.list_succ == 0].sort_values("list_total", ascending=False)

Dates  list_total  \
JRNL_ID                                                                  
1_54942  [article_not_in_Scihub, article_not_in_Scihub,...          37   
1_27702  [no_date_found, no_date_found, no_date_found, ...          22   
1_28313  [no_date_found, no_date_found, no_date_found, ...          22   
1_64747  [no_date_found, no_date_found, no_date_found, ...          22   
1_45401  [article_not_in_Scihub, article_not_in_Scihub,...          20   
...                                                    ...         ...   
1_50636                                    [no_date_found]           1   
1_50663                            [article_not_in_Scihub]           1   
1_50715                                    [no_date_found]           1   
1_35456                                    [no_date_found]           1   
1_26742                                    [no_date_found]           1   

         list_succ  list_diff  
JRNL_ID                        
1_54942          0         37  
1_27702          0         22  
1_28313          0         22  
1_64747          0         22  
1_45401          0         20  
...            ...        ...  
1_50636          0          1  
1_50663          0          1  
1_50715          0          1  
1_35456          0          1  
1_26742          0          1  

[1976 rows x 4 columns]

In [171]:
q1_anlys[q1_anlys.list_succ == 0].list_total.value_counts().sort_index()

1     955
2     427
3     230
4     100
5      68
6      51
7      36
8      28
9      15
10     17
11     15
12      6
13      6
14      7
15      3
16      1
17      2
18      2
19      1
20      2
22      3
37      1
Name: list_total, dtype: int64

In [176]:
q1_jrnl_df[q1_jrnl_df.list_diff > 0].sort_values("list_diff",ascending=False)

ISSN  Year  \
ID                                    
1_25758  (10959203, 00368075)  2011   
1_37840  (14764687, 00280836)  2010   
1_25759  (10959203, 00368075)  2018   
1_54942  (18060374, 21790302)  2019   
1_63003  (20955138, 2053714X)  2015   
...                       ...   ...   
1_34098  (14653923, 00905992)  2015   
1_34060  (14653648, 02681153)  2015   
1_34059  (14653648, 02681153)  2013   
1_34051  (14653419, 13557858)  2015   
1_66912  (8756758X, 14602695)  2016   

                                                     Title  \
ID                                                           
1_25758                                            Science   
1_37840                                             Nature   
1_25759                                            Science   
1_54942              Periodico Tche Quimica (discontinued)   
1_63003                            National Science Review   
...                                                    ...   
1_34098                               Nationalities Papers   
1_34060                          Health Education Research   
1_34059                          Health Education Research   
1_34051                               Ethnicity and Health   
1_66912  Fatigue and Fracture of Engineering Materials ...   

                                                  Sbj_Area  Total_Docs  \
ID                                                                       
1_25758  {'Arts and Humanities': 1, 'Multidisciplinary'...        2093   
1_37840                           {'Multidisciplinary': 1}        2464   
1_25759  {'Arts and Humanities': 1, 'Multidisciplinary'...        2943   
1_54942  {'Multidisciplinary': 1, 'Health Professions': 1}         265   
1_63003                           {'Multidisciplinary': 1}          75   
...                                                    ...         ...   
1_34098   {'Social Sciences': 2, 'Arts and Humanities': 1}          47   
1_34060                             {'Social Sciences': 1}          50   
1_34059              {'Social Sciences': 1, 'Medicine': 1}          93   
1_34051   {'Arts and Humanities': 1, 'Social Sciences': 1}          36   
1_66912         {'Materials Science': 1, 'Engineering': 2}         120   

                Country            Region  \
ID                                          
1_25758   United States  Northern America   
1_37840  United Kingdom    Western Europe   
1_25759   United States  Northern America   
1_54942          Brazil     Latin America   
1_63003  United Kingdom    Western Europe   
...                 ...               ...   
1_34098  United Kingdom    Western Europe   
1_34060  United Kingdom    Western Europe   
1_34059  United Kingdom    Western Europe   
1_34051  United Kingdom    Western Europe   
1_66912  United Kingdom    Western Europe   

                                                 Publisher  \
ID                                                           
1_25758  American Association for the Advancement of Sc...   
1_37840                            Nature Publishing Group   
1_25759  American Association for the Advancement of Sc...   
1_54942                                 Tche Quimica Group   
1_63003                            Oxford University Press   
...                                                    ...   
1_34098                                          Routledge   
1_34060                            Oxford University Press   
1_34059                            Oxford University Press   
1_34051                                          Routledge   
1_66912                     Wiley-Blackwell Publishing Ltd   

                     Coverage     SJR  ...  Citable Docs. (3years)  \
ID                                     ...                           
1_25758  1880-1881, 1883-2020  14,238  ...                    4336   
1_37840             1869-2020  16,465  ...                    4587   
1_25759  1880-1881, 1883-2020  13,251  ...                    4537   
1_54

NOTE: IT IS IMPORTANT TO NOTE THAT, DURING PARTIAL FIX OR OTHER RUNS, IT IS REEEEAAAAALLLYYY CRUCIAL TO CONSIDER "FAILED" DOIs!!!!!

To ensure unnnecessary duplicate record retrival for already failed artcls!

In [186]:
set(q1_meta_fe.DOI.tolist()).difference(set(q1_date_df.index))

29774

In [182]:
q1_jrnl_df.fillna(0)[q1_jrnl_df.fillna(0).Required_Count < q1_jrnl_df.fillna(0).list_total]

ISSN  Year                        Title  \
ID                                                                 
1_49971  (16512316, 16506073)  2014  Cognitive Behaviour Therapy   

                  Sbj_Area  Total_Docs Country          Region  \
ID                                                               
1_49971  {'Psychology': 1}          31  Sweden  Western Europe   

                     Publisher   Coverage    SJR  ...  Citable Docs. (3years)  \
ID                                                ...                           
1_49971  Taylor and Francis AS  2002-2020  1,391  ...                      87   

         Cites / Doc. (2years)  Ref. / Doc.  CR_Count  SH_Results  \
ID                                                                  
1_49971                   2,39        45,77        38        True   

        Current_Count Required_Count  list_total  list_succ  list_diff  
ID                                                                      
1_49971           1.0              5        10.0        9.0        1.0  

[1 rows x 24 columns]

In [188]:
len(q1_meta_fe)

138184

In [187]:
len(q1_date_fe)

108345

In [189]:
138184-108345

29839

In [190]:
29839-29774

65